### PR TITLE
Add fullName to resource class and keep resource consistent across lg resources

### LINKED ISSUE #4152: Keep LG resource IDs consistent
Currently, there are two modes of ID generation, one is the top level of lg resource, which would return the `filePath` as its id, but for child lg resource,(which is imported by top lg),  `filename` is the id. It would cause redundant parse.

How to fix:
Make the id consistent, make them all filename or make them all filepath.

### PR DESCRIPTION
**Pull request by boydc2014**
Keep LG resource IDs consistent

Fixes #4152

The issue previously is, assuming such a structure:
1.lg import 2.lg
The id of 1.lg is set to the file full path, -> xxx/yy/1.lg,
But the id of 2.lg is set to the file name, -> 2.lg
which is not consistent

What this pr did:

1. change `ImportResolverDelegate` from 
`(string sourceId, string resourceId) => (string content, string id)` to `(LGResource resource, string resourceId) => LGResource` and `LGResource` contains Id, Content and FullName properties.

2. Add `ParseResource` function which a more generic LGResource into LGParser to replace ParseText  (`ParseText` would be deprecated and marked as [Obsolete])

3. Add `TemplateEngineLanguageGenerator(Resource resource,  resourceMapping)` constructor,
`filePath` and `lg text` format constructor would be deprecated in the next version. Mark them with [Obsolete]

4. Move `FullName` property from `FileResource` to abstract class `Resource`

5. Use `lgResource.Id` as the id of the unified templates to call

6. Add `Source` property in `Templates` which means the source of current lg file. (which would be absolute file path for file resource.)

7. Adjust corresponding tests.

### CODE REVIEW

**Comment by boydc2014** [comment]
A squash of https://github.com/microsoft/botbuilder-dotnet/pull/4155

**Comment by tomlm** [comment]
>         public static ImportResolverDelegate ResourceExplorerResolver(string locale, Dictionary<string, IList<Resource>> resourceMapping)

The signature for this is that it returns an Id, and you are returning the fullname.  Maybe you should change to return (content, resource) ?

---
Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/LanguageGeneratorManager.cs:56 in ac723d9. [](commit_id = ac723d93c0eafad61f1801ddaeaf818bfe7304ba, deletion_comment = False)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/LanguageGeneratorManager.cs [hunk 1]
```diff
@@ -70,7 +70,7 @@ public static ImportResolverDelegate ResourceExplorerResolver(string locale, Dic
                 else
                 {
                     var content = resource.ReadTextAsync().GetAwaiter().GetResult();
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/LanguageGeneratorManager.cs:72]: >.ReadTextAsync().GetAwaiter().GetResult(); [](start = 42, length = 42)

eeesh

**Comment by boydc2014** [comment]
> > ```
> >     public static ImportResolverDelegate ResourceExplorerResolver(string locale, Dictionary<string, IList<Resource>> resourceMapping)
> > ```
> 
> The signature for this is that it returns an Id, and you are returning the fullname. Maybe you should change to return (content, resource) ?
> 
> Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/LanguageGeneratorManager.cs:56 in [ac723d9](https://github.com/microsoft/botbuilder-dotnet/commit/ac723d93c0eafad61f1801ddaeaf818bfe7304ba). [](commit_id = [ac723d9](https://github.com/microsoft/botbuilder-dotnet/commit/ac723d93c0eafad61f1801ddaeaf818bfe7304ba), deletion_comment = False)

we should use id and return id, lg resources should follow the convention in resource explorer that use fileName as id. FullName should only be used later when debugger are interested, not supposed to be used here. I will check with @Danieladu tomorrow

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs [hunk 2]
```diff
@@ -63,9 +47,9 @@ public TemplateEngineLanguageGenerator(string lgText, string id, Dictionary<stri
         public TemplateEngineLanguageGenerator(string filePath, Dictionary<string, IList<Resource>> resourceMapping)
         {
             filePath = PathUtils.NormalizePath(filePath);
-            this.Id = Path.GetFileName(filePath);
+            this.Id = filePath;
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs:50]: >this.Id = filePath; [](start = 12, length = 19)

no, I think I would keep Id as the resource.id
Add a new property for the fullname.

File: tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Templates.Tests/LGGeneratorTests.cs [hunk 3]
```diff
@@ -47,8 +47,7 @@ public async Task TestMultiLangImport()
             // use LG file as entrance
             var lgResourceGroup = LGResourceLoader.GroupByLocale(resourceExplorer);
 
-            var resource = resourceExplorer.GetResource("a.en-US.lg") as FileResource;
-            var generator = new TemplateEngineLanguageGenerator(resource.FullName, lgResourceGroup);
+            var generator = new TemplateEngineLanguageGenerator(resourceExplorer.GetResource("a.en-US.lg").FullName, lgResourceGroup);
```
Comment (tomlm) [inline tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Templates.Tests/LGGeneratorTests.cs:50]: >TemplateEngineLanguageGenerator [](start = 32, length = 31)

Isn't this an opportunity to make the LG fully resource based?  Why not add in a constructor which takes a resource?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/LanguageGeneratorManager.cs (hunk 1 above)
Comment (boydc2014) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/LanguageGeneratorManager.cs]: is there a better way to read async data and feed into a sync method?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs (hunk 2 above)
Comment (boydc2014) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs]: updated

File: tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Templates.Tests/LGGeneratorTests.cs (hunk 3 above)
Comment (boydc2014) [reply tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Templates.Tests/LGGeneratorTests.cs]: updated, add a new method for this

**Comment by mrivera-ms** [comment]
>         public static ImportResolverDelegate ResourceExplorerResolver(string locale, Dictionary<string, IList<Resource>> resourceMapping)

consider making this method async.

---
Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/LanguageGeneratorManager.cs:56 in a52c5c9. [](commit_id = a52c5c9906bf61478de64207f47bb7bb8691e991, deletion_comment = False)

**Comment by Danieladu** [comment]
> > ```
> >     public static ImportResolverDelegate ResourceExplorerResolver(string locale, Dictionary<string, IList<Resource>> resourceMapping)
> > ```
> 
> Consider making this method async.
> 
> Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/LanguageGeneratorManager.cs:56 in [a52c5c9](https://github.com/microsoft/botbuilder-dotnet/commit/a52c5c9906bf61478de64207f47bb7bb8691e991). [](commit_id = [a52c5c9](https://github.com/microsoft/botbuilder-dotnet/commit/a52c5c9906bf61478de64207f47bb7bb8691e991), deletion_comment = False)

We are considering change the ImportResolverDelegate from `(sourceId, resourceId ) => (content, id)` to `(sourceId, resourceId ) => LGResource`, Maybe `async (sourceId, resourceId ) => LGResource` is more resonable

**Comment by boydc2014** [comment]
> > > ```
> > >     public static ImportResolverDelegate ResourceExplorerResolver(string locale, Dictionary<string, IList<Resource>> resourceMapping)
> > > ```
> > 
> > 
> > Consider making this method async.
> > Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/LanguageGeneratorManager.cs:56 in [a52c5c9](https://github.com/microsoft/botbuilder-dotnet/commit/a52c5c9906bf61478de64207f47bb7bb8691e991). [](commit_id = [a52c5c9](https://github.com/microsoft/botbuilder-dotnet/commit/a52c5c9906bf61478de64207f47bb7bb8691e991), deletion_comment = False)
> 
> We are considering change the ImportResolverDelegate from `(sourceId, resourceId ) => (content, id)` to `(sourceId, resourceId ) => LGResource`, Maybe `async (sourceId, resourceId ) => LGResource` is more resonable

async (LGResource src, string target) => Resource looks better to me, src should be a resource, instead of plain string, which gives more more information on source to leverage when resolving an import statement

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs [hunk 4]
```diff
@@ -47,27 +47,47 @@ public TemplateEngineLanguageGenerator(LanguageGeneration.Templates engine = nul
         /// <param name="lgText">lg template text.</param>
         /// <param name="id">optional label for the source of the templates (used for labeling source of template errors).</param>
         /// <param name="resourceMapping">template resource loader delegate (locale) -> <see cref="ImportResolverDelegate"/>.</param>
+        [Obsolete("This method will soon be deprecated. Use Resource as the first parameter instead.")]
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs:50]: use ParseResource method instead

**Comment by boydc2014** [comment]
@tomlm this one is good for a second look now, we updated the description to reflect the latest change, in which mainly two changes,
1. we upgrade the importResolver to be LGResource  based
2. we upgrade parseText (depreciating this one) to parseResource which will take an generic LGResource 

@mrivera-ms thanks for the suggestions of make this async, we will follow up on this maybe as a second PR, because that will impact also the caller side like parseFile, parseResource, we will have to rename those to align the rules.

**Comment by mrivera-ms** [comment]
@tomlm , could you please take a look at this one? Dong has made the changes requested.

**Comment by boydc2014** [comment]
@tomlm ping~

**Pull request merged**